services/httpd/pprof.go

```diff
@@ -68,60 +62,173 @@ type prof struct {
+	//
+
+	// last() returns either the last item from a slice of strings or an empty
+	// string if the supplied slice is empty or nill.
+	last := func(s []string) string {
+		if len(s) == 0 {
+			return ""
 		}
-		allProfs = append([]*prof{profile}, allProfs...) // CPU profile first.
+		return s[len(s)-1]
 	}
 
-	var (
-		resp bytes.Buffer // Temporary buffer for entire archive.
-		buf  bytes.Buffer // Temporary buffer for each profile/query result.
-	)
+	// if trace exsits as a form value, add it to the profiles slice with the
+	// decoded duration.
+	//
+	// Requests for a trace should look like:
+	//
+	//  ?trace=10s
+	//
+	if vals, exists := r.Form["trace"]; exists {
+		// parse the duration encoded in the last "trace" value supplied.
+		duration, err := time.ParseDuration(last(vals))
+
+		// If we can't parse the duration or if the users supplies a negative
+		// number, then set it to a reasonable default of 10s.
+		//
+		// This allows for stuff like ?trace=true -- which isn't recommened but
+		// makes the code do something reasonable.
+		//
+		// If someone wants to supply a trace duration of 0s I guess that's their
+		// perogative but lets ensure that it is positive.
+		if defaultDuration := time.Second * 10; err != nil || duration < 0 {
+			duration = defaultDuration
+		}
 
-	gz := gzip.NewWriter(&resp)
-	tw := tar.NewWriter(gz)
+		// Trace files can get big.  Lets clamp the maximum trace duration to 45s.
+		if maxDuration := time.Second * 45; duration > maxDuration {
+			duration = maxDuration
+		}
 
+		profiles = append(profiles, prof{"trace", duration})
+	}
+
+	// Capturing CPU profiles is a little tricker.  The preferred way to send the
+	// the cpu profile duration is via the supplied "cpu" variable's value.
+	//
+	// The duration should be encoded as a Go duration that can be parsed by
+	// time.ParseDuration().
+	//
+	// In the past users were encouraged to assign any value to cpu and provide
+	// the duration in a separate "seconds" value.
+	//
+	// The code below handles both -- first it attempts to use the old method
+	// which would look like:
+	//
+	//    ?cpu=foobar&seconds=10
+	//
+	// Then it attempts to ascertain the duration provided with:
+	//
+	//    ?cpu=10s
+	//
+	// This preserves backwards compatibility with any tools that have been
+	// written to gather profiles.
+	//
+	if vals, exists := r.Form["cpu"]; exists {
+		duration := time.Second * 30
+		val := last(vals)
+		if seconds, exists := r.Form["seconds"]; exists {
+			s, _ := strconv.ParseInt(last(seconds), 10, 64)
+			duration = time.Second * time.Duration(s)
+		} else if val != "" {
+			// see if the value of cpu is a duration like:  cpu=10s
+			duration, _ = time.ParseDuration(val)
+		}
+
+		// ensure that the duration is positive
+		if defaultDuration := time.Second * 30; duration < 0 {
+			duration = defaultDuration
+		}
+
+		// ensure that profile duration is less than 1 minute
+		if maxDuration := time.Minute * 1; duration > maxDuration {
```

Comment:
Was there a max duration for profiles before? I'm not seeing it at a glance. If not, 1 minute might be too restrictive. Five minutes maybe? Although, I'm not sure we should have a max.

Reply:
there wasn't a max.  the standard was 30 seconds.  i figured some kind of restriction is better than none.  but maybe if you're poking this endpoint you know what you're doing.

Reply:
okay -- removed.  i think it *is* reasonable to limit trace duration to 45 seconds.  i figure 30 seconds is probably more than anyone would want and could digest since the traces are so dense.  then i added 50% to allow for a little extra -- so 45 seconds of trace.